cardano-cli/test/cardano-cli-test/Test/Cli/Hash.hs

```diff
@@ -0,0 +1,211 @@
+              , "--expected-hash"
+              , exampleAnchorDataHash
+              ]
+      )
+
+-- | Execute me with:
+-- @cabal test cardano-cli-test --test-options '-p "/check anchor data hash from ipfs uri/"'@
+hprop_check_anchor_data_hash_from_ipfs_uri :: Property
+hprop_check_anchor_data_hash_from_ipfs_uri =
+  propertyOnce $ do
+    let relativeUrl = ["ipfs", exampleAchorDataIpfsHash]
+    serveFileWhile
+      relativeUrl
+      exampleAnchorDataPath
+      ( \port -> do
+          env <- H.evalIO getEnvironment
+          result <-
+            execDetailCfgCardanoCLI
+              H.defaultExecConfig
+                { H.execConfigEnv =
+                    Last $
+                      Just
+                        ( ( "IPFS_GATEWAY_URI"
+                          , "http://localhost:" ++ show port ++ "/"
+                          )
+                            : env
+                        )
+                }
+              [ "hash"
+              , "anchor-data"
+              , "--url"
+              , "ipfs://" ++ exampleAchorDataIpfsHash
+              , "--expected-hash"
+              , exampleAnchorDataHash
+              ]
+          case result of
+            (ExitFailure _, _, stderr) -> do
+              H.note_ stderr
+              failure
+            (ExitSuccess, _, _) -> success
+      )
+
+-- | Takes a relative url (as a list of segments), a file path, and an action, and it serves
+-- the file in the url provided in a random free port that is passed as a parameter to the
+-- action. After the action returns, it shuts down the server. It returns the result of the
+-- action. It also ensures the server is shut down even if the action throws an exception.
+serveFileWhile
+  :: (MonadBaseControl IO m, MonadTest m, MonadIO m, HasCallStack)
+  => [String]
+  -- ^ Relative URL where the file will be served.
+  -- Each element is a segment of the URL.
+  -> FilePath
+  -- ^ File path for the file to serve
+  -> (Int -> m a)
+  -- ^ Action to run while the file is being served.
+  -- It receives the port the server is listening on
+  -> m a
+serveFileWhile relativeUrl filePath action =
+  bracket
+    -- Server setup (resource acquisition)
+    ( do
+        -- Get the port the server is listening on
+        (port, socket) <- H.evalIO openFreePort
+        -- Serve the file
+        let app :: Request -> (Response -> IO ResponseReceived) -> IO ResponseReceived
+            app req respond = do
+              let path = T.unpack <$> pathInfo req
+              if path == relativeUrl
+                -- Status -> ResponseHeaders -> FilePath -> Maybe FilePart -> Response
+                then respond $ responseFile status200 [("Content-Type", "text/plain")] filePath Nothing
+                else respond $ responseLBS status404 [("Content-Type", "text/plain")] "404 - Not Found"
+
+        -- Run server asynchronously in a separate thread
+        void $ H.evalIO $ forkOS $ runSettingsSocket defaultSettings socket app
```

Comment:
Why `forkOS` and not `forkIO`?

Reply:
I initially wrote `forkIO` but then I changed it trying to fix an issue (I read a disclaimer with system calls in some systems having trouble). I think it turned out **not** to be the cause of the problem, but I left it like that. Probably safer, cause I am not sure where syscalls are happening.

Reply:
It doesn't matter much, but it just looks odd. I'm fine with either.